akka-stream/src/main/scala/akka/stream/impl/fusing/FlatMapPrefix.scala

```diff
@@ -0,0 +1,169 @@
+      }
+
+      override def onUpstreamFailure(ex: Throwable): Unit = {
+        subSource match {
+          case OptionVal.Some(s) => s.fail(ex)
+          case OptionVal.None    =>
+            //flow won't be materialized, so we have to complete the future with a failure indicating this
+            matPromise.failure(new NeverMaterializedException(ex))
+            super.onUpstreamFailure(ex)
+        }
+      }
+
+      override def onPull(): Unit = {
+        subSink match {
+          case OptionVal.Some(s) =>
+            //delegate to subSink
+            s.pull()
+          case OptionVal.None if accumulated.size < n =>
+            pull(in)
+          case OptionVal.None if accumulated.size == n =>
+            //corner case for n = 0, can be handled in FlowOps
+            materializeFlow()
+        }
+      }
+
+      override def onDownstreamFinish(cause: Throwable): Unit = {
+        subSink match {
+          case OptionVal.None    => downstreamCause = OptionVal.Some(cause)
+          case OptionVal.Some(s) => s.cancel(cause)
+        }
+      }
+
+      def materializeFlow(): Unit =
+        try {
+          val prefix = accumulated.toVector
+          accumulated.clear()
+          subSource = OptionVal.Some(new SubSourceOutlet[In](s"${this}.subSource"))
+          val OptionVal.Some(theSubSource) = subSource
+          theSubSource.setHandler {
+            new OutHandler {
+              override def onPull(): Unit = {
+                if (!isClosed(in) && !hasBeenPulled(in)) {
+                  pull(in)
+                }
+              }
+
+              override def onDownstreamFinish(cause: Throwable): Unit = {
+                if (!isClosed(in)) {
+                  cancel(in, cause)
+                }
+              }
+            }
+          }
+          subSink = OptionVal.Some(new SubSinkInlet[Out](s"${this}.subSink"))
+          val OptionVal.Some(theSubSink) = subSink
+          theSubSink.setHandler {
+            new InHandler {
+              override def onPush(): Unit = {
+                push(out, theSubSink.grab())
+              }
+
+              override def onUpstreamFinish(): Unit = {
+                complete(out)
+              }
+
+              override def onUpstreamFailure(ex: Throwable): Unit = {
+                fail(out, ex)
+              }
+            }
+          }
+          val matVal = try {
+            val flow = f(prefix)
+            val runnableGraph = Source.fromGraph(subSource.get.source).viaMat(flow)(Keep.right).to(subSink.get.sink)
+            interpreter.subFusingMaterializer.materialize(runnableGraph)
+          } catch {
+            case NonFatal(ex) =>
+              matPromise.failure(new NeverMaterializedException(ex))
+              subSource = OptionVal.None
+              subSink = OptionVal.None
+              throw ex
+          }
+          matPromise.success(matVal)
+
+          //in case downstream was closed
+          if (downstreamCause.isDefined) {
```

Comment:
Maybe use `downstreamCause.foreach` (if that's defined on OptionVal, didn't check)?